### ISSUE TITLE
Lua filter round-trip tests "double free or corruption" error

### ISSUE THREAD

**Issue by jgm** [jgm/pandoc#6674]
We're seeing this in linux CI:
```
 Lua
3592
    Lua filters
3593
      inline elements can be round-tripped through the lua stack:    OK (0.10s)
3594
        +++ OK, passed 20 tests.
3595
      block elements can be round-tripped through the lua stack:     OK (0.14s)
3596
        +++ OK, passed 20 tests.
3597
      meta blocks can be round-tripped through the lua stack:        double free or corruption (!prev)
3598
Test suite test-pandoc: FAIL
```
https://github.com/jgm/pandoc/runs/1098511351?check_suite_focus=true

@tarleb any ideas?

**Comment by jgm** [comment]
This may be related to the recent pandoc-types changes?

**Comment by tarleb** [comment]
Yikes. This could be due to changes in HsLua 1.2: that version contains a number of low-level changes, including improvements to the way functions are exposed to Lua. This might take a while to debug; we should probably set the hslua < 1.2 constraint again.

**Comment by tarleb** [comment]
OK, I have tightened the version constraints for hslua again.

The error is a mystery to me. I'm unable to reproduce it locally despite running the tests many times (same compiler version, same OS as far as I can tell).

**Comment by tarleb** [comment]
Unfortunately, I can still provoke this error and similar failures by running the test suite with the number of QuickCheckTests to a large (four digit) value. Most failures are related to double-free issues, but there's the occasional round-trip failure without a crash. I assumed those validation failures would be repeatable by setting the quickcheck seed, but that not the case.

At least the error is rare, it occurs once every 100-500 CI runs by my estimate.

I'm reopening this and will have to dig deeper. I'll also loosen the hslua version constraints again, as that did not help.

**Comment by jgm** [comment]
Here's another instance of it
https://github.com/jgm/pandoc/runs/1579669659?check_suite_focus=true#step:7:3909

**Comment by tarleb** [comment]
So far I double-checked my memory allocating Haskell code, but didn't find anything. Guess I'll have to learn how to use Valgrind to really track this down.

**Comment by tarleb** [comment]
Using Valgrind, I now can see some memory handling problems, but don't quite know how to proceed. My interpretation of the valgrind output is that there is something going wrong in Lua's memory allocator, but that would be very surprising.

```
      meta blocks can be round-tripped through the lua stack:     ==670637== Thread 5:                                                                                                                             
==670637== Invalid write of size 8                                                                                                                                                                                 
==670637==    at 0x817DE35: lua_pushlstring (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                          
==670637==    by 0x814A4FE: ??? (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                      
==670637==  Address 0x10749fc0 is 0 bytes after a block of size 544 alloc'd                                                                                                                                        
==670637==    at 0xB538FAF: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)                                                                                                      
==670637==    by 0x817F5B0: l_alloc (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                  
==670637==    by 0x818A4A9: luaM_realloc_ (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                            
==670637==    by 0x81847D6: luaD_reallocstack (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)
==670637==    by 0x818704F: propagatemark (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                            
==670637==    by 0x81877E7: singlestep (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)   
==670637==    by 0x8187F95: luaC_step (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)        
==670637==    by 0x816A740: ??? (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)          
==670637==                                                                                                                                                                                                         
==670637== Invalid write of size 4                                                                                                                                                                                 
==670637==    at 0x817DE46: lua_pushlstring (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)  
==670637==    by 0x814A4FE: ??? (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)
==670637==  Address 0x10749fc8 is 8 bytes after a block of size 544 alloc'd                                                                                                                                        
==670637==    at 0xB538FAF: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)                                                                                                      
==670637==    by 0x817F5B0: l_alloc (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                  
==670637==    by 0x818A4A9: luaM_realloc_ (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                            
==670637==    by 0x81847D6: luaD_reallocstack (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                        
==670637==    by 0x818704F: propagatemark (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                            
==670637==    by 0x81877E7: singlestep (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                               
==670637==    by 0x8187F95: luaC_step (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                
==670637==    by 0x816A740: ??? (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                      
==670637==                                                                                                                                                                                                         
==670637== Invalid read of size 16                                                                                                                                                                                 
==670637==    at 0x819486E: luaH_setint (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                              
==670637==    by 0x817E75E: lua_rawseti (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                              
==670637==    by 0x8168802: hsluazm1zi3zi0zma1e01b238d09bf15a3c38922ede20bfea87e8a1d8cd7a77e19cbfcd16199e6e6_ForeignziLuaziTypesziPushable_zdfPushableZLz2cUZR2_info (in /home/albert/projects/pandoc/pandoc/dist-n
ewstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                                                                                                    
==670637==  Address 0x10749fc0 is 0 bytes after a block of size 544 alloc'd                                                                                                                                        
==670637==    at 0xB538FAF: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)                                                                                                      
==670637==    by 0x817F5B0: l_alloc (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                  
==670637==    by 0x818A4A9: luaM_realloc_ (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                            
==670637==    by 0x81847D6: luaD_reallocstack (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                        
==670637==    by 0x818704F: propagatemark (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                            
==670637==    by 0x81877E7: singlestep (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                               
==670637==    by 0x8187F95: luaC_step (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                
==670637==    by 0x816A740: ??? (in /home/albert/projects/pandoc/pandoc/dist-newstyle/build/x86_64-linux/ghc-8.6.5/pandoc-2.11.4/t/test-pandoc/build/test-pandoc/test-pandoc)                                      
==670637== 
```

**Comment by jgm** [comment]
Right, it's more likely to be an issue in hslua, but I don't know anything about hslua so it's hard to comment further!  Is there any way to figure out how to reproduce this deterministically?

**Comment by tarleb** [comment]
Not sure yet: in the rare case that the tests produced a wrong result, but didn't crash, re-running the tests with the same seed did not give the same error. I therefore assume that it's not deterministic, but I'm still going to write a separate program to debug this further.

**Comment by jgm** [comment]
Looks like the problems are with pushing strings to the lua stack, correct?
That is at least helpful information. Maybe use QuickCheck to exercise hslua on random Strings, Texts, ByteStrings, and see if this pops up again?  It would be nice to know whether the issue is in hslua or in pandoc specifically.

**Comment by tarleb** [comment]
I wasn't able yet to construct a more minimal test case, but the I have a suspicion and, from that, a working work-around:

It appears that Lua's and Haskell's memory management get in each others way. I don't know the reason for this, as I'm still not aware of any problematic uses of the FFI in hslua (based on Lua and Haskell docs). Regardless, the problem was no longer reproducible after compiling pandoc with `--constraints 'hslua -allow-unsafe-gc'`. Here's how I had documented this flag:

> Allow optimizations which make Lua's garbage collection potentially unsafe; Haskell finalizers must be handled with extreme care.

Disabling the flag forces Haskell to do extra bookkeeping work before and after calling out to C/Lua. Maybe this *is* the real problem, or maybe it just hides a race condition, but it definitely solves the issue. The performance hit from this is a slow-down of about ~50%. Given how rare the issue is, we could think about setting this flag for testing only. It doesn't feel quite right though.

**Comment by tarleb** [comment]
Small addendum: while I wasn't able to reproduce the double free issue when using safe GC, I did see one case of "Exception: heap overflow" which I had never seen before. But at least that error doesn't crash the test suite.

**Comment by tarleb** [comment]
Correction on the slowdown: and average filter runtime rises to somewhere between 200% and 300% of the optimized version.

**Comment by jgm** [comment]
Well, that's progress at least!  I'm not sure what to suggest about setting the option.  That's a big slowdown, and setting an option for testing that we don't use for the release seems maybe a bad idea.  I think I'd rather keep things as they are for now; hopefully we might eventually isolate the issue more precisely.

**Comment by tarleb** [comment]
I figured it out. The issue is that the hslua function which retrieves strings from the Lua stack is not strict enough. The effect is that Haskell is sometimes trying to read a C string when it's already too late and the string has been gc'ed by Lua. I think this can be fixed in hslua, but in the meantime we can just increase strictness in our unmarshaling functions to force evaluation. I've prepared a patch and didn't see the error reappearing despite running an excessive amount of tests. I can't say for sure if the issue has been eliminated completely, but at least it is now even less likely to occur.
